src/main/java/com/mixpanel/android/mpmetrics/MixpanelNotificationRouteActivity.java

```diff
@@ -0,0 +1,95 @@
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class MixpanelNotificationRouteActivity extends Activity {
+
+    protected final String LOGTAG = "MixpanelAPI.MixpanelNotificationRouteActivity";
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+
+        Intent routeIntent = getIntent();
+        Bundle extras = routeIntent.getExtras();
+
+        if (null == routeIntent) {
+            MPLog.i(LOGTAG, "Notification route activity given null intent.");
+            return;
+        }
+
+        trackAction(routeIntent);
+
+        Intent notificationIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(routeIntent.getExtras().getCharSequence("uri").toString()));
+        if (!extras.getBoolean("sticky")) {
+            cancelNotification(extras);
+        }
+        startActivity(notificationIntent);
+    }
+
+    private void cancelNotification(Bundle extras) {
+        int notificationId = extras.getInt("notificationId");
+        NotificationManager notificationManager = (NotificationManager) getApplicationContext().getSystemService(getApplicationContext().NOTIFICATION_SERVICE);
+        notificationManager.cancel(notificationId);
+    }
+
+    protected void trackAction(Intent routeIntent) {
+        Bundle intentExtras = routeIntent.getExtras();
+
+        CharSequence actionIdChars = intentExtras.getCharSequence("actionId");
+        if (null == actionIdChars) {
+            MPLog.i(LOGTAG, "Notification action click logged with no actionId.");
+            return;
+        }
+
+        CharSequence uriChars = intentExtras.getCharSequence("uri");
+        if (null == uriChars) {
+            MPLog.i(LOGTAG, "Notification action click logged with no uri.");
+            return;
+        }
+
+        CharSequence messageIdChars = intentExtras.getCharSequence("messageId");
+        if (null == messageIdChars) {
+            MPLog.i(LOGTAG, "Notification action click logged with no messageId.");
+            return;
+        }
+
+        CharSequence campaignIdChars = intentExtras.getCharSequence("campaignId");
+        if (null == campaignIdChars) {
+            MPLog.i(LOGTAG, "Notification action click logged with no campaignId.");
+            return;
+        }
+
+        final String actionId = actionIdChars.toString();
+        final String uri = uriChars.toString();
+        final String messageId = messageIdChars.toString();
+        final String campaignId = campaignIdChars.toString();
+
+        MixpanelAPI.allInstances(new MixpanelAPI.InstanceProcessor() {
+            @Override
+            public void process(MixpanelAPI api) {
+                JSONObject pushProps = new JSONObject();
+
+                try {
+                    pushProps.put("actionId", actionId);
+                    pushProps.put("uri", uri);
+                    pushProps.put("messageId", messageId);
+                    pushProps.put("campaignId", campaignId);
```

Comment:
we should use snake_case for these properties to match all of our other message tracking events: `action_id`, `message_id`, `campaign_id`. Also should we call it `button_id` instead of `action_id`? I'm trying to think through whether the `action` abstraction makes sense but I don't know.

Reply:
https://github.com/mixpanel/mixpanel-android/pull/642/files fixed here